File: Algorithms/p203_Remove_Linked_List_Elements/p203_Remove_Linked_List_Elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode** pp = &head;
        while(*pp){
            if((*pp)->val==val){
                // save and free *pp
                *pp = (*pp)->next;
            }else{
                pp = &(*pp)->next;
            }
        }
        return head;
    }
};

File: Algorithms/p203_Remove_Linked_List_Elements/p203_Remove_Linked_List_Elements.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        head0 = ListNode(None)
        head0.next = head
        p = head0
        while p.next:
            if p.next.val==val:
                p.next = p.next.next
            else:
                p = p.next
        return head0.next
